CMSgov/mint-app
# EASI-1988/plan list by collaborator

# EASI-1988

## Changes and Description

- Plan Collaborators used to get model plans
- Initializer of model becomes the model Lead Collaborator
- Unable to delete the last model lead from the plan collaborator.
- Add basic unit test for collaborator.



## How to test this change
1. `run scripts/dev test:go` 
     a. Note that the test pass
     b. I had to disable deleting model plan from the plan milestones unit test, as it would violate a foreign key constraint

2. `Run  scripts/dev db:clean && scripts/dev db:seed`
    a. Note that there are records for the seeded model plans
    
3. Run the following GQL scripts to note new functionality

```gql
mutation createModelPlan {
  createModelPlan(input: {modelName: "My Custom Model", id: "9b5db64b-8d42-4b91-a601-d6f9732b5298"
status: PLAN_DRAFT}) {
    id
    createdBy
    collaborators {
      id
      euaUserID
      teamRole
      fullName
    }
  }
}
```


attempt to delete the plan collaborator that was created when the model plan was created. Update the id to be from the query above

```gql
mutation deletePlanCollaborator {
  deletePlanCollaborator(input: {
        id: "",
        modelPlanID: "9b5db64b-8d42-4b91-a601-d6f9732b5298",
        euaUserID: "MINT",
        teamRole: MODEL_LEAD,
        fullName: "mint Doe",

  }) {
      id
      euaUserID
      teamRole
      fullName
  }
}
```

Note that that should give you an error

add another model team lead with this query

```gql
mutation createPlanCollaborator {
  createPlanCollaborator(input: {
    id: "a76dc7b7-331d-4876-8ce2-926fbd06249c"
    modelPlanID: "9b5db64b-8d42-4b91-a601-d6f9732b5298"
    euaUserID: "WJZZ"
    fullName:"Steven Wade"

    teamRole: MODEL_LEAD
  }) {
    id
    modelPlanID
    euaUserID
    fullName

    teamRole
  }
}
```

Try to delete the initial collaborator again. It should now be allowed since there is another Model Lead Listed


Now try to update the newest collaborator to a role that isn't model lead. It won't be allowed since there is now only 1 model_lead

```gql
mutation updatePlanCollaborator {
  updatePlanCollaborator(input: {
    id: "a76dc7b7-331d-4876-8ce2-926fbd06249c"
    modelPlanID: "9b5db64b-8d42-4b91-a601-d6f9732b5298"
    euaUserID: "WJZZ"
    fullName:"Steven Wade"

    teamRole: LEADERSHIP
  }) {
    id
    modelPlanID
    euaUserID
    fullName

    teamRole
  }
}
```
## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
